rfcs/0048-on-conflict-clause-in-create-table-statement.md

```diff
@@ -0,0 +1,121 @@
+---
+feature: ON CONFLICT Clause in Create Table Statement
+authors:
+  - "st1page"
+start_date: "2023/2/1"
+---
+
+# ON CONFLICT Clause in Create Table Statement
+
+## Summary
+
+allow user to declare the conflict behavior when the newly inserted row break the unique constraint of primary key.
+
+## Motivation
+
+Currently, for a table with primary key, the later operations will always overwrite the previous ones when the new operation [0017-DML.md](0017-DML.md)
+But users' requirement are various here.
+I will give some different behavior here. Notice that here the conflict behavior is not just happens when an insert DML statement happening but also in a table with external connector.
+
+- **Overwrite**
+  The most important needs of using the Overwrite behavior is to express "Upsert" operation. Some connector can not give the old value of the update operation and we need to overwrite the old value with the same pk in storage.
+
+- **Ignore**
+  The Ignore behavior make user can define a "append-only table with primary key" which maintain the primary constrain and also provide a append-only changes to downstream with good performance. This is good to express a table with an At-least-once Delivery connector where a same record can be replayed multiple times to break the primary key constrain.
+  Notice: the specific grammar for this has not designed(e.g. should we expose append-only table to user?) but the situation is the same.
+
+- **Partial Update**
+  The ON CONFLICT make users can do partial update on the table.
+
+  ```SQL
+    CREATE table t(k int primary key, int a, int b) 
+    ON CONFLICT DO UPDATE SET (a,b) = (
+      COALESCE(EXCLUDED.a, a),
+      COALESCE(EXCLUDED.b, b));
+  ```
+  
+  This feature will be useful when user use wide-table schema data model.
+
+- **Express Reduce Function or UDAF?**
+  In fact, ON CONFLICT Clause can make user be able to do more things. User can even construct a streaming Map-reduce job with that where the primary key define the map fields and ON_CONFLICT clause define the reduce logic.
+  For example.
+
+  ```SQL
+    CREATE table t(k int primary key, int cnt) 
+    ON CONFLICT DO UPDATE SET cnt = t.cnt + EXCLUDED.cnt
```

Comment:
Why we don't use aggregation call here:
```
select k, sum(cnt) from t;
```

Reply:
It is just a trivial example here to show that this grammar can let user declare their own aggregation logic with SQL. And the loigic could be more complex.